docs/advanced_topics_checkpointing.rst

```diff
@@ -0,0 +1,89 @@
+Checkpointing
+=============
+
+.. currentmodule:: nvidia.dali
+
+Checkpointing is a feature in DALI which allows you to save the current state of the pipeline to a file.
+Then, you can restore the pipeline from a saved checkpoint and the new pipeline will produce exactly the same outputs as the old one would.
+It is particularly useful for long-running training jobs which are likely to be interrupted.
+
+A checkpoint of DALI pipeline contains information about states of all random number generators used in the pipeline and about the progress of each reader.
+
+Checkpointing API
+-----------------
+
+Enabling checkpointing
+~~~~~~~~~~~~~~~~~~~~~~
+
+To enable checkpointing, set ``enable_checkpointing=True`` when creating a pipeline.
+With this option enabled, DALI will track the state of each operator, allowing you to save it on demand.
+Enabling checkpointing shouldn't have any impact on the performance.
+
+.. code-block:: python
+
+  @pipeline_def(..., enable_checkpointing=True)
+  def pipeline():
+      ...
+
+  p = pipeline()
+  p.build()
+
+
+.. note::
+    Readers with ``shuffle_after_epoch=True`` might produce different (but still equally random) output if checkpointing is enabled.
```

Comment:
Just a suggestion, but I'd make it a bit more concrete (i.e. shuffle, not produce outputs). At the same time, I would avoid going into the details on randomness quality. The fact that we had such conernes seems to me like a detail of design/implementation steps, but, as lon as we do not degrade it, I would not bother user with that at all.

```suggestion
    Readers with ``shuffle_after_epoch=True`` might shuffle samples differently if checkpointing is enabled.
```

Reply:
I wanted to make sure that the user won't understand this as "shuffling is not random if checkpointing is enabled"

Reply:
But I guess "shuffle samples differently" implies that it's still random